flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointTest.java

```diff
@@ -0,0 +1,491 @@
+
+            StopWithSavepoint sws =
+                    createStopWithSavepoint(ctx, mockExecutionGraph, savepointFuture);
+            ctx.setStopWithSavepoint(sws);
+            ctx.setExpectFinished(assertNonNull());
+
+            savepointFuture.complete(SAVEPOINT_PATH);
+            mockExecutionGraph.completeTerminationFuture(JobStatus.FINISHED);
+            ctx.triggerScheduledExecutors();
+
+            assertThat(sws.getOperationFuture().get(), is(SAVEPOINT_PATH));
+        }
+    }
+
+    @Test
+    public void testJobFailed() throws Exception {
+        try (MockStopWithSavepointContext ctx = new MockStopWithSavepointContext()) {
+            StateTrackingMockExecutionGraph mockExecutionGraph =
+                    new StateTrackingMockExecutionGraph();
+            StopWithSavepoint sws = createStopWithSavepoint(ctx, mockExecutionGraph);
+            ctx.setStopWithSavepoint(sws);
+            ctx.setHowToHandleFailure(Executing.FailureResult::canNotRestart);
+
+            ctx.setExpectFailing(
+                    failingArguments -> {
+                        assertThat(
+                                failingArguments.getExecutionGraph().getState(),
+                                is(JobStatus.FAILED));
+                        assertThat(
+                                failingArguments.getFailureCause(),
+                                containsCause(FlinkException.class));
+                    });
+
+            // fail job:
+            mockExecutionGraph.completeTerminationFuture(JobStatus.FAILED);
+            ctx.triggerScheduledExecutors();
+
+            assertThat(sws.getOperationFuture().isCompletedExceptionally(), is(true));
+        }
+    }
+
+    @Test
+    public void testJobFinishedBeforeSavepointFuture() throws Exception {
+        try (MockStopWithSavepointContext ctx = new MockStopWithSavepointContext()) {
+            StateTrackingMockExecutionGraph mockExecutionGraph =
+                    new StateTrackingMockExecutionGraph();
+            CompletableFuture<String> savepointFuture = new CompletableFuture<>();
+
+            StopWithSavepoint sws =
+                    createStopWithSavepoint(ctx, mockExecutionGraph, savepointFuture);
+            ctx.setStopWithSavepoint(sws);
+            ctx.setExpectFinished(assertNonNull());
+
+            mockExecutionGraph.completeTerminationFuture(JobStatus.FINISHED);
+
+            savepointFuture.complete(SAVEPOINT_PATH);
+            ctx.triggerScheduledExecutors();
+
+            assertThat(sws.getOperationFuture().get(), is(SAVEPOINT_PATH));
+        }
+    }
+
+    @Test
+    public void testTransitiontoCancellingOnCancel() throws Exception {
+        try (MockStopWithSavepointContext ctx = new MockStopWithSavepointContext()) {
+            StopWithSavepoint sws = createStopWithSavepoint(ctx);
+            ctx.setStopWithSavepoint(sws);
+            ctx.setExpectCancelling(assertNonNull());
+
+            sws.cancel();
```

Comment:
We should check that the operation has been failed.

Reply:
Fair enough. We have tested this with `testExceptionalFutureCompletionOnLeaveWhileWaitingOnSavepointCompletion`